File: 2025/C/answers/java/Solution.java
import java.util.*;
import java.io.*;
import java.math.BigInteger;

public class Solution {
    static class Cleaner {
        int h, l, u;
        long m;
        public Cleaner(int h, int l, int u, long m) {
            this.h = h; this.l = l; this.u = u; this.m = m;
        }
    }

    static class Rational implements Comparable<Rational> {
        int num, den;
        public Rational(int num, int den) {
            this.num = num; this.den = den;
        }
        @Override
        public int compareTo(Rational other) {
            long p1 = (long) num * other.den;
            long p2 = (long) other.num * den;
            return Long.compare(p1, p2);
        }
        @Override
        public boolean equals(Object obj) {
            if (!(obj instanceof Rational)) return false;
            Rational other = (Rational) obj;
            return (long) num * other.den == (long) other.num * den;
        }
    }

    static class Event implements Comparable<Event> {
        Rational coord;
        int type; // 0 start, 1 end
        int id;
        public Event(Rational coord, int type, int id) {
            this.coord = coord; this.type = type; this.id = id;
        }
        @Override
        public int compareTo(Event other) {
            int cmp = this.coord.compareTo(other.coord);
            if (cmp != 0) return cmp;
            return Integer.compare(this.type, other.type);
        }
    }

    static long[] bitCnt;
    static long[] bitSum;
    static int nh;
    static int[] distinctH;

    static void update(int idx, long valM, long valMH) {
        for (; idx <= nh; idx += idx & -idx) {
            bitCnt[idx] += valM;
            bitSum[idx] += valMH;
        }
    }

    static long getMinSumH(long s) {
        int idx = 0;
        long currentCnt = 0;
        long currentSum = 0;
        int maxBit = Integer.highestOneBit(nh);
        for (int i = maxBit; i > 0; i >>= 1) {
            int nextIdx = idx + i;
            if (nextIdx <= nh && currentCnt + bitCnt[nextIdx] < s) {
                idx = nextIdx;
                currentCnt += bitCnt[idx];
                currentSum += bitSum[idx];
            }
        }
        long remaining = s - currentCnt;
        currentSum += remaining * distinctH[idx];
        return currentSum;
    }

    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt()) return;
        int n = sc.nextInt();
        long s = sc.nextLong();
        long k = sc.nextLong();

        List<Cleaner> cleaners = new ArrayList<>();
        List<Event> events = new ArrayList<>();
        Set<Integer> hSet = new TreeSet<>();

        for (int i = 0; i < n; i++) {
            int h = sc.nextInt();
            int l = sc.nextInt();
            int u = sc.nextInt();
            long m = k / h;
            if (m >= 1) {
                int id = cleaners.size();
                cleaners.add(new Cleaner(h, l, u, m));
                events.add(new Event(new Rational(l, h), 0, id));
                events.add(new Event(new Rational(u, h), 1, id));
                hSet.add(h);
            }
        }

        if (cleaners.isEmpty()) {
            System.out.println("*");
            return;
        }

        distinctH = new int[hSet.size()];
        int hPtr = 0;
        for (int h : hSet) distinctH[hPtr++] = h;
        nh = distinctH.length;

        bitCnt = new long[nh + 1];
        bitSum = new long[nh + 1];

        Collections.sort(events);

        long bestNum = -1, bestDen = 1;

        int i = 0;
        while (i < events.size()) {
            Rational currentCoord = events.get(i).coord;
            int j = i;
            while (j < events.size() && events.get(j).coord.equals(currentCoord)) {
                if (events.get(j).type == 0) {
                    Cleaner c = cleaners.get(events.get(j).id);
                    int hIdx = Arrays.binarySearch(distinctH, c.h) + 1;
                    update(hIdx, c.m, c.m * c.h);
                }
                j++;
            }

            long totalCnt = 0;
            for (int idx = nh; idx > 0; idx -= idx & -idx) totalCnt += bitCnt[idx];

            if (totalCnt >= s) {
                long sumH = getMinSumH(s);
                long curNum = (long) currentCoord.num * sumH;
                long curDen = currentCoord.den;
                if (bestNum == -1 || compareFractions(curNum, curDen, bestNum, bestDen) < 0) {
                    bestNum = curNum;
                    bestDen = curDen;
                }
            }

            j = i;
            while (j < events.size() && events.get(j).coord.equals(currentCoord)) {
                if (events.get(j).type == 1) {
                    Cleaner c = cleaners.get(events.get(j).id);
                    int hIdx = Arrays.binarySearch(distinctH, c.h) + 1;
                    update(hIdx, -c.m, -c.m * c.h);
                }
                j++;
            }
            i = j;
        }

        if (bestNum == -1) {
            System.out.println("*");
        } else {
            long common = gcd(bestNum, bestDen);
            System.out.println((bestNum / common) + " " + (bestDen / common));
        }
    }

    static int compareFractions(long n1, long d1, long n2, long d2) {
        BigInteger p1 = BigInteger.valueOf(n1).multiply(BigInteger.valueOf(d2));
        BigInteger p2 = BigInteger.valueOf(n2).multiply(BigInteger.valueOf(d1));
        return p1.compareTo(p2);
    }

    static long gcd(long a, long b) {
        return b == 0 ? a : gcd(b, a % b);
    }
}

File: 2025/C/answers/kt/Solution.kt
import java.util.*
import java.math.BigInteger

class Cleaner(val h: Int, val l: Int, val u: Int, val m: Long)

class Rational(val num: Int, val den: Int) : Comparable<Rational> {
    override fun compareTo(other: Rational): Int {
        val p1 = num.toLong() * other.den
        val p2 = other.num.toLong() * den
        return p1.compareTo(p2)
    }

    override fun equals(other: Any?): Boolean {
        if (other !is Rational) return false
        return num.toLong() * other.den == other.num.toLong() * den
    }

    override fun hashCode(): Int {
        return 0
    }
}

class Event(val coord: Rational, val type: Int, val id: Int) : Comparable<Event> {
    override fun compareTo(other: Event): Int {
        val cmp = coord.compareTo(other.coord)
        if (cmp != 0) return cmp
        return type.compareTo(other.type)
    }
}

var bitCnt = LongArray(0)
var bitSum = LongArray(0)
var nh = 0
var distinctH = IntArray(0)

fun update(index: Int, valM: Long, valMH: Long) {
    var idx = index
    while (idx <= nh) {
        bitCnt[idx] += valM
        bitSum[idx] += valMH
        idx += idx and -idx
    }
}

fun getMinSumH(s: Long): Long {
    var idx = 0
    var currentCnt = 0L
    var currentSum = 0L
    var i = Integer.highestOneBit(nh)
    while (i > 0) {
        val nextIdx = idx + i
        if (nextIdx <= nh && currentCnt + bitCnt[nextIdx] < s) {
            idx = nextIdx
            currentCnt += bitCnt[idx]
            currentSum += bitSum[idx]
        }
        i = i shr 1
    }
    val remaining = s - currentCnt
    currentSum += remaining * distinctH[idx]
    return currentSum
}

fun gcd(a: Long, b: Long): Long = if (b == 0L) a else gcd(b, a % b)

fun compareFractions(n1: Long, d1: Long, n2: Long, d2: Long): Int {
    val p1 = BigInteger.valueOf(n1).multiply(BigInteger.valueOf(d2))
    val p2 = BigInteger.valueOf(n2).multiply(BigInteger.valueOf(d1))
    return p1.compareTo(p2)
}

fun main() {
    val sc = Scanner(System.`in`)
    if (!sc.hasNextInt()) return
    val n = sc.nextInt()
    val s = sc.nextLong()
    val k = sc.nextLong()

    val cleaners = mutableListOf<Cleaner>()
    val events = mutableListOf<Event>()
    val hSet = TreeSet<Int>()

    for (i in 0 until n) {
        val h = sc.nextInt()
        val l = sc.nextInt()
        val u = sc.nextInt()
        val m = k / h
        if (m >= 1) {
            val id = cleaners.size
            cleaners.add(Cleaner(h, l, u, m))
            events.add(Event(Rational(l, h), 0, id))
            events.add(Event(Rational(u, h), 1, id))
            hSet.add(h)
        }
    }

    if (cleaners.isEmpty()) {
        println("*")
        return
    }

    distinctH = hSet.toIntArray()
    nh = distinctH.size
    bitCnt = LongArray(nh + 1)
    bitSum = LongArray(nh + 1)

    Collections.sort(events)

    var bestNum = -1L
    var bestDen = 1L

    var i = 0
    while (i < events.size) {
        val currentCoord = events[i].coord
        var j = i
        while (j < events.size && events[j].coord == currentCoord) {
            if (events[j].type == 0) {
                val c = cleaners[events[j].id]
                val hIdx = Arrays.binarySearch(distinctH, c.h) + 1
                update(hIdx, c.m, c.m * c.h)
            }
            j++
        }

        var totalCnt = 0L
        var idx = nh
        while (idx > 0) {
            totalCnt += bitCnt[idx]
            idx -= idx and -idx
        }

        if (totalCnt >= s) {
            val sumH = getMinSumH(s)
            val curNum = currentCoord.num.toLong() * sumH
            val curDen = currentCoord.den.toLong()
            if (bestNum == -1L || compareFractions(curNum, curDen, bestNum, bestDen) < 0) {
                bestNum = curNum
                bestDen = curDen
            }
        }

        j = i
        while (j < events.size && events[j].coord == currentCoord) {
            if (events[j].type == 1) {
                val c = cleaners[events[j].id]
                val hIdx = Arrays.binarySearch(distinctH, c.h) + 1
                update(hIdx, -c.m, -c.m * c.h)
            }
            j++
        }
        i = j
    }

    if (bestNum == -1L) {
        println("*")
    } else {
        val common = gcd(bestNum, bestDen)
        println("${bestNum / common} ${bestDen / common}")
    }
}
